File: Semestru 4/MPP/lab/.net/lab3/lab3/Networking/utils/AbstractServer.cs
using log4net;
using System.Net.Sockets;
using System.Net;

namespace Networking.utils
{
	public abstract class AbstractServer
	{
		private TcpListener Server;
		private String Host;
		private int Port;

		private static readonly ILog Log = LogManager.GetLogger(typeof(AbstractServer));
		public AbstractServer(String host, int port)
		{
			this.Host = host;
			this.Port = port;
		}
		public void Start()
		{
			IPAddress adr = IPAddress.Parse(Host);
			IPEndPoint ep = new IPEndPoint(adr, Port);
			Server = new TcpListener(ep);
			Server.Start();
			while (true)
			{
				Log.Debug("Waiting for clients ...");
				TcpClient client = Server.AcceptTcpClient();
				Log.Debug("Client connected ...");
				processRequest(client);
			}
		}

		public void Stop()
		{
			if (Server != null)
				Server.Stop();
		}

		public abstract void processRequest(TcpClient client);

	}
}

File: Semestru 4/MPP/lab/java/lab3/Network/src/main/java/org/example/utils/AbstractServer.java
package org.example.utils;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;
import java.rmi.ServerException;

public abstract class AbstractServer {
    private int port;
    private ServerSocket server=null;
    private static Logger logger = LogManager.getLogger(AbstractServer.class);
    public AbstractServer( int port){
        this.port=port;
    }

    public void start() throws ServerException {
        try{
            server=new ServerSocket(port);
            while(true){
                logger.info("Waiting for clients ...");
                Socket client=server.accept();
                logger.info("Client connected ...");
                //new Thread(() -> processRequest(client)).start();
                processRequest(client);
            }
        } catch (IOException e) {
            throw new ServerException("Starting server errror ",e);
        }finally {
            stop();
        }
    }

    protected abstract  void processRequest(Socket client);
    public void stop() throws ServerException {
        try {
            server.close();
        } catch (IOException e) {
            throw new ServerException("Closing server error ", e);
        }
    }
}
